File: lib/render/item.dart
import 'dart:io';

import '../util.dart';
import '../world.dart';
import 'render.dart';

class Item extends Render {

  late ItemModel model;

  Item(super.location, super.name, this.model);

  Item.json(super.location, super.json) : super.json();

  Item.resource(super.location, super.name, Map<String, dynamic> json) {
    resource(json);
  }

  @override
  void json(Map<String, dynamic> json) {
    super.json(json);
    model = ItemModel.json(json);
  }

  @override
  Map<String, dynamic> toJson() => super.toJson()..addAll(model.toJson());

  void resource(Map<String, dynamic> json) {
    model = ItemModel.resource(json);
  }

  Future<void> save() async {
    File file = loader.dataFile("item", location, "json");
    await file.create(recursive: true);
    await file.writeAsString(encoder.convert(toJson()).replaceAllMapped(
        RegExp(r'\[\s*([\d.,\s]+)\s*\]'), (match) => '[${match.group(1)!.replaceAll(RegExp(r'\s+'), ' ').trim()}]'
    ));
  }
}

class ItemModel implements JsonMappable<Map<String, dynamic>> {

  Display display = Display();
  List<Cube> cubes = [];

  ItemModel(this.cubes);

  ItemModel.json(Map<String, dynamic> json) {
    this.json(json);
  }

  ItemModel.resource(Map<String, dynamic> json, {Map<String, String> pTextures = const {}}) {
    resource(json, pTextures);
  }

  void json(Map<String, dynamic> json) {
    display = Display.json(json["display"]);
    cubes = List.generate(json["cubes"].length, (index) => Cube.json(json["cubes"][index]));
  }

  Map<String, dynamic> toJson() => {
    "display": display.toJson(),
    "cubes": List.generate(cubes.length, (index) => cubes[index].toJson())
  };

  void resource(Map<String, dynamic> json, Map<String, String> pTextures) {
    Map<String, String> textures = {
      if(json["textures"] != null)
        for(String key in json["textures"].keys)
          key: json["textures"][key][0] == "#" ? pTextures[json["textures"][key].substring(1)] ?? json["textures"][json["textures"][key].substring(1)]! : json["textures"][key],
      for(String key in pTextures.keys)
        key: pTextures[key]!
    };
    if(json["elements"] != null) {
      cubes = List.generate(json["elements"].length, (index) => Cube.resource(json["elements"][index], textures));
    }
    if(json["parent"] != null) {
      Location parentLoc = Location.json(json["parent"]);
      if(parentLoc == Location.minecraft("builtin/entity")) {
        print("Item model skipped, it is from \"builtin/entity\"");
      } else if(parentLoc == Location.minecraft("item/generated")) {
        cubes.addAll([
          for(String layer in textures.keys)
            if(layer.indexOf("layer") == 0)
              Cube(
                  dimension: Dimension(Pos.zero, Size(16, 16, 0)),
                  faces: {"north": Texture(loader.dataFile("textures", Location.json(textures[layer]!), "png"))..save(loader.resource("textures", Location.json(textures[layer]!), "png")!)}
              )
        ]);
        display + itemGenerated;
      } else if(parentLoc == Location.minecraft("item/handheld")) {
        cubes.addAll([
          for(String layer in textures.keys)
            if(layer.indexOf("layer") == 0)
              Cube(
                  dimension: Dimension(Pos.zero, Size(16, 16, 0)),
                  faces: {"north": Texture(loader.dataFile("textures", Location.json(textures[layer]!), "png"))..save(loader.resource("textures", Location.json(textures[layer]!), "png")!)}
              )
        ]);
        display + itemHandheld;
      } else {
        ItemModel parent = ItemModel.resource(loader.resourceJson("models", Location.json(json["parent"])), pTextures: textures);
        cubes.addAll(parent.cubes);
        display + parent.display;
      }
    }
  }
}

final Display itemGenerated = Display(
    gui: DisplayConfig(rotation: Rotation.zero, translation: Pos.zero, scale: Size(1, 1, 1)),
    fixed: DisplayConfig(rotation: Rotation(0, 180, 0), translation: Pos.zero, scale: Size(1, 1, 1))
);

final Display itemHandheld = Display(
    gui: DisplayConfig(rotation: Rotation.zero, translation: Pos.zero, scale: Size(1, 1, 1)),
    fixed: DisplayConfig(rotation: Rotation(0, 180, 0), translation: Pos.zero, scale: Size(1, 1, 1))
);

class Display implements JsonMappable<Map<String, dynamic>> {

  late DisplayConfig? gui;
  late DisplayConfig? fixed;

  Display({this.gui, this.fixed});

  Display.json(Map<String, dynamic> json) {
    this.json(json);
  }

  @override
  void json(Map<String, dynamic> json) {
    if(json["gui"] != null) {
      gui = DisplayConfig.json(json["gui"]);
    }
    if(json["fixed"] != null) {
      fixed = DisplayConfig.json(json["fixed"]);
    }
  }

  @override
  Map<String, dynamic> toJson() => {
    if(gui != null) "gui": gui!.toJson(),
    if(fixed != null) "fixed": fixed!.toJson()
  };

  void operator +(Display other) {
    if(gui == null) {
      gui = other.gui;
    }
    if(fixed == null) {
      fixed = other.fixed;
    }
  }
}

class DisplayConfig implements JsonMappable<Map<String, dynamic>> {

  late Rotation rotation;
  late Pos translation;
  late Size scale;

  DisplayConfig({required this.rotation, required this.translation, required this.scale});

  DisplayConfig.json(Map<String, dynamic> json) {
    this.json(json);
  }

  @override
  void json(Map<String, dynamic> json) {
    rotation = json["rotation"];
    translation = json["translation"];
    scale = json["scale"];
  }

  @override
  Map<String, dynamic> toJson() => {
    "rotation": rotation,
    "translation": translation,
    "scale": scale
  };
}

File: src/minecraft/objects/item.ts
import { Cube, RenderObject, Texture } from "../render.js"
import { Pos3D, Rotation3D, Size3D } from "../../world/world3D.js"
import { MinecraftObject } from "./object.js"
import { loader, Location } from "../loader.js"
import path from "path"

export class Item extends MinecraftObject {

  model: ItemModel

  constructor(location: Location, name: string, model: ItemModel) {
    super(location, name)
    this.model = model
  }

  static resource(location: Location, name: string, json: any): Item {
    return new Item(location, name, ItemModel.resource(json))
  }

  static fromJson(location: Location, json: any): Item {
    return new Item(location, json.name, ItemModel.fromJson(json.model))
  }

  toJson(): any {
    return {
      name: this.name,
      model: this.model.toJson(this.location)
    }
  }

  defaultRender(): RenderObject {
    return this.model.render
  }

  get path(): string {
    return loader.dataFile("item", this.location, "json")
  }

  renderToSave(): Record<string, RenderObject> {
    return Object.fromEntries([
      [path.join(loader.renderDir, 'items', `${this.location.toDir()}.json`), this.model.render]
    ])
  }
}

export class ItemModel {

  constructor(public render: RenderObject, public display: Display = new Display()) {
    this.render = render
  }

  static fromJson(json: any): ItemModel {
    return new ItemModel(new RenderObject(json.cubes.map((cube: any) => Cube.fromJson(cube))), Display.fromJson(json.display))
  }

  static resource(json: any, pTextures: Record<string, string> = {}): ItemModel {
    const model = new ItemModel(new RenderObject())

    const textures: Record<string, string> = {
      ...(json.textures ? Object.fromEntries(
        Object.keys(json.textures).map((key) => {
          const value = json.textures[key]
          return [
            key,
            (value as string)[0] === '#' ? pTextures[(value as string).substring(1)] || json.textures[(value as string).substring(1)] : value
          ]
        })
      ) : {}),
      ...pTextures
    }

    if (json.elements) {
      model.render = json.elements.map((element: any) => Cube.resource(element, textures))
    }

    if (json.parent) {
      const parentLoc = Location.fromJson(json.parent)
      if (parentLoc.equals(Location.minecraft('builtin/entity'))) {
        console.log(`Item model of ${parentLoc} skipped, it is from 'builtin/entity'`)
      } else if (parentLoc.equals(Location.minecraft('item/generated'))) {
        model.render.cubes.push(...Object.keys(textures)
          .filter(layer => layer.indexOf('layer') === 0)
          .map(layer => {
            const cube = new Cube(
              new Pos3D(0.5, 0, 0.5),
              new Size3D(1, 0, 1)
            )
            const location = Location.fromJson(textures[layer].replace('/', '\\'))
            const texture = new Texture(location)
            texture.save(loader.renderFile('textures', location, 'png'), loader.resource('textures', Location.fromJson(textures[layer]), 'png')!)
            cube.faces[0] = texture
            return cube
          })
        )
        model.display.add(Display.itemGenerated)
      } else if (parentLoc.equals(Location.minecraft('item/handheld'))) {
        model.render.cubes.push(...Object.keys(textures)
          .filter(layer => layer.indexOf('layer') === 0)
          .map(layer => {
            const cube = new Cube(
              new Pos3D(0.5, 0, 0.5),
              new Size3D(1, 0, 1)
            )
            const location = Location.fromJson(textures[layer].replace('/', '\\'))
            const texture = new Texture(location)
            texture.save(loader.renderFile('textures', location, 'png'), loader.resource('textures', Location.fromJson(textures[layer]), 'png')!)
            cube.faces[0] = texture
            return cube
          })
        )
        model.display.add(Display.itemHandheld)
      } else {
        const parent = ItemModel.resource(loader.resourceJson('models', Location.fromJson(json.parent)), textures)
        model.render.cubes.push(...parent.render.cubes)
        model.display.add(parent.display)
      }
    }
    return model
  }

  toJson(location: Location): any {
    return {
      display: this.display.toJson(),
      render: path.join(loader.renderDir, 'items', `${location.toDir()}.json`)
    }
  }
}

export class DisplayConfig {

  constructor(readonly rotation: Rotation3D, readonly translation: Pos3D, readonly scale: Size3D) { }

  static fromJson(json: Record<string, any>): DisplayConfig {
    return new DisplayConfig(
      json.rotation,
      json.translation,
      json.scale
    )
  }

  toJson(): any {
    return {
      rotation: this.rotation,
      translation: this.translation,
      scale: this.scale
    }
  }
}

export class Display {

  static readonly itemGenerated = new Display(
    new DisplayConfig(Rotation3D.north(), new Pos3D(0, 0, 0), new Size3D(1, 1, 1)),
    new DisplayConfig(Rotation3D.south(), new Pos3D(0, 0, 0), new Size3D(1, 1, 1))
  )

  static readonly itemHandheld = new Display(
    new DisplayConfig(Rotation3D.north(), new Pos3D(0, 0, 0), new Size3D(1, 1, 1)),
    new DisplayConfig(Rotation3D.south(), new Pos3D(0, 0, 0), new Size3D(1, 1, 1))
  )

  constructor(public gui?: DisplayConfig, public fixed?: DisplayConfig) { }

  static fromJson(json: Record<string, any>): Display {
    return new Display(DisplayConfig.fromJson(json.gui), DisplayConfig.fromJson(json.fixed))
  }

  toJson(): any {
    return {
      gui: this.gui?.toJson(),
      fixed: this.fixed?.toJson()
    }
  }

  add(other: Display): void {
    if (!this.gui) {
      this.gui = other.gui
    }
    if (!this.fixed) {
      this.fixed = other.fixed
    }
  }
}